### ISSUE TITLE
Unable to ping pods between each other over non-default service record for k8s 1.21 cluster

### ISSUE THREAD

**Issue by thehejik** [rancher/rancher#33085]
**What kind of request is this (question/bug/enhancement/feature request):** bug

**Steps to reproduce (least amount of steps as possible):**
TL;DR - pods in one namespace cannot be pinged between each other over custom dns record pointing to a workload - pings are passing over records created automatically when you deploy a deployment. The same testcase described below is working fine on custom RKE1 cluster with k8s `v1.20.7-rancher1-1` deployed in the same local cluster.

Reproducer:
* deploy custom RKE1 cluster with k8s `v1.21.1-rancher1-1` on single node with all roles under local cluster running `rancher/rancher:master-head 7d7dd99` (on RKE1 with k8s v1.20.6)
* on the downstream cluster deploy `busybox1` deployment in `default` namespace, enter `busybox` container image end specify Command: `/bin/sh -c "while true; do echo blah; sleep 1; done"` - do not expose any port
* clone the `busybox1` deployment and name it `busybox2`
* create `arecord` custom service in `default` namespace under Default project->Resources->Workload->Service Discovery, select `Resolves To:`->`One or more workloads` bullet and point the dns entry to `busybox2` deployment/workload. 
* Execute shell in `busybox1-xxx` pod and perform: `ping arecord` - it will return error `ping: bad address 'arecord'`

**Result:**
``` bash
/ # ping arecord
ping: bad address 'arecord'

/ # ping arecord.default.svc.cluster.local
ping: bad address 'arecord.default.svc.cluster.local'

/ # ping busybox1
PING busybox1 (10.42.3.71): 56 data bytes
64 bytes from 10.42.3.71: seq=0 ttl=64 time=0.052 ms

/ # ping busybox2
PING busybox2 (10.42.3.72): 56 data bytes
64 bytes from 10.42.3.72: seq=0 ttl=64 time=0.026 ms

/ # ping busybox1.default.svc.cluster.local
PING busybox1.default.svc.cluster.local (10.42.3.71): 56 data bytes
64 bytes from 10.42.3.71: seq=0 ttl=63 time=0.054 ms
```

**Other details that may be helpful:**
* The very same test case is working fine on RKE1 cluster with k8s v1.20.6
* default DNS entries created automatically for deployments are working
* seems to be similar to v1.21 k3s issue https://github.com/rancher/rancher/issues/32329

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): `rancher/rancher:master-head 7d7dd99`
- Installation option (single install/HA): single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): any
- Kubernetes version (use `kubectl version`): `v1.21.1-rancher1-1`

**Comment by kinarashah** [comment]
Feature gate `IPv6DualStack` for IPv4/IPv6 dual-stack networking is enabled by default for 1.21 kubernetes cluster. But cluster needs to meet a few requirements and configurations before it works correctly. 

So we're going to disable the feature flag by default in Rancher k8s 1.21.

**Comment by kinarashah** [comment]
Available to test with RKE `v1.3.0-rc6` and rancher master-head once build passes, https://drone-publish.rancher.io/rancher/rancher/5709

**Comment by thehejik** [comment]
Blocked by https://github.com/rancher/rancher/issues/32801

**Comment by kinarashah** [comment]
Available to test once build passes https://drone-publish.rancher.io/rancher/rancher/5719

**Comment by thehejik** [comment]
**Test report:**
Tested on `rancher/rancher:master-head 5715b8a` on RKE1 custom cluster with `v1.21.1-rancher2-1`.
* For any workload the corresponding services with deployment name are created automatically
* I was able ping pods from each other over automatically created services for the workload by using `ping <deployment name>`
* I was able ping pods from each other over custom service as described in summary

Only problem I found is that services which are created automatically for the workload are not getting removed once the workload is removed, I opened new issue https://github.com/rancher/rancher/issues/33178

This is working ok. Close.

**Comment by kinarashah** [comment]
Disabling Feature gate isn't the ideal way to resolve this, since it is enabled by default in upstream. 
We need to take a look into what is UI/API passing to create service and figure out what we're doing different than services created by default in Rancher or a v1.service created using kubectl.

**Comment by superseb** [comment]
This is caused by this:

```
Headless Services without selector
For Headless Services without selectors and without .spec.ipFamilyPolicy explicitly set, the .spec.ipFamilyPolicy field defaults to RequireDualStack.
```

**Comment by thehejik** [comment]
Note: once this will be out there shoudn't be `--feature-gates=IPv6DualStack=false` arg used for `kube-apiserver` process on nodes (on RKE1 with v1.21 and eg. Canal CNI)

**Comment by thehejik** [comment]
There is no sign of `--feature-gates` arg for `kube-apiserver` process on single-node RKE1 cluster with k8s `v1.21.1-rancher2-1` provisioned on local `rancher/rancher:master-head cd623b5` anymore.

The services are created automatically for a deployment defined in Ember UI and I was able to perform ping from one deployment/pod to another by `ping <anameofdeployment>`

But custom services created in EmberUI which are pointing to a workload (used `Resolves to`: `One or more workloads`) are not working again, a new issue was opened in meantime https://github.com/rancher/rancher/issues/33755 for this.

BTW if I create a service with `Resolves to`: `Alias of another DNS record's value` the ping is working.

**Comment by superseb** [comment]
The root cause is still the same as https://github.com/rancher/rancher/issues/33085#issuecomment-867835847, although I'm seeing in the current requests for creating a service, that `ipFamilyPolicy` is not empty but actually set to the default in Kubernetes (`RequireDualStack`). Instead of fixing this in the backend, I would add the configurable field `ipFamilyPolicy` to the service creation page and have it default to `SingleStack` so it works by default and if people want to create dual-stack service records, they can change it.

If we do want to fix this in the backend, we need to detect if the cluster is properly configured for dual-stack, which will then determine if the service record is going to be `SingleStack` or `RequireDualStack`.

**Comment by superseb** [comment]
@gaktive Is this something you can discuss if this could go into the UI?

**Comment by vincent99** [comment]
I don't know the whole story here, but the UI making up a default that is different than the actual API's default (and us turning off features that are in by default?) is fairly unusual...

**Comment by superseb** [comment]
The feature gate `IPv6DualStack` was enabled by default in v1.21, even though no dual stack configuration exists. We follow upstream, but having that feature gate enabled, causes headless services to default to dualstack while there is no dualstack configuration. So as it needs to be configurable anyway (user can choose what service he/she wants to create), we can add it and default to singlestack and advanced users (I would mark dualstack as advanced currently), and it will solve the issue.

We can also take it a step further back and disable the feature gate but that was the first fix and reverted as we needed to follow upstream.

**Comment by brandond** [comment]
This isn't a viable fix for K3s and RKE2. We are going out of our way to allow for dual-stack operation; telling people to rebuild their cluster with dual-stack disabled to run Rancher isn't going to fly.

**Comment by thehejik** [comment]
I tried according to the reproducer in summary on `rancher/rancher:master-head 94bca99` and RKE1 downstream cluster with `k8s v1.21.1-rancher3-1` and the custom services are still not reachable. I could reach only pods over their automatically created services by `ping <deployment name>`.

Reopening.

**Comment by sowmyav27** [comment]
@thehejik Please validate with a newer version of mater-head

**Comment by anupama2501** [comment]
Verified on master-head 76b0695

- Manually created two workloads
- Created a Service discovery for one of the workloads
- Pinged the SD from the other workload and works fine

**Comment by anupama2501** [comment]
Blocked by https://github.com/rancher/rancher/issues/33755 for other validations.

**Comment by superseb** [comment]
Anything created through the Ember UI is validated as working. The part that is currently stuck is that we have automated tests that use `type: dnsRecord` that has been removed from using in the Ember UI since a while (https://github.com/rancher/ui/commit/1ed161e60c2282e7e18a058838b58410a205aecc) that create Service recources that are not going through the Store that we modified to make this work. Currently looking into the dnsRecord functionality.

**Comment by superseb** [comment]
Available with `c5b24bff8f4c29b71de49a783dc2764953b3116a`

**Comment by anupama2501** [comment]
Verified on master-head f04ae9f

DNS resolution for the service discovery pods is working as expected 
Validation tests run via automation have passed.
